### PR TITLE
Playlists

### PR DESCRIPTION
**Pull request by minikin**
**What is in this PR:**
In-memory implementation of:
- Create a Playlist
- Add audio to Playlist on Audio Player
- Edit Playlist
- Search for Playlist
- Reorder of audio files in a Playlist
- Various bug fixes and improvements

**What is not in this PR**
- Integration with Firebase

**Known Bugs**
- <s>Add to Playlist screen is presented behind AudioPlayer (Overlay)</s>

### CODE REVIEW

**Comment by minikin** [comment]
Pls, review @7flash @JasonHighVibe

**Comment by minikin** [comment]
@7flash Pls, review.

File: ios/Podfile.lock [hunk 1]
```diff
@@ -214,6 +214,16 @@ PODS:
   - DKPhotoGallery/Resource (0.0.14):
     - SDWebImage
     - SDWebImageFLPlugin
+  - FBSDKCoreKit (5.15.1):
```
Comment (7flash) [inline ios/Podfile.lock:217]: Podfile.lock should not be included in commit but rather added to .gitignore or reverted

File: lib/modules/audio_player/audio_player_page.dart [hunk 2]
```diff
@@ -24,10 +28,13 @@ class AudioPlayerPageState
 
   @override
   Widget build(BuildContext context) {
-    return Observer(
-      builder: (_) => controller.isMinimized
-          ? SafeArea(child: _minimizedAudioPlayer())
-          : Material(child: _fullScreenAudioPlayer()),
+    return Offstage(
+      offstage: AppModule.audioPlayerPageIsOffstage,
```
Comment (7flash) [inline lib/modules/audio_player/audio_player_page.dart:32]: Why?

File: lib/modules/audio_player/audio_player_page.dart [hunk 3]
```diff
@@ -370,6 +375,20 @@ class AudioPlayerPageState
...
                             },
+                          ),
+                          IconButton(
+                            icon: const Icon(
+                              Icons.playlist_add,
+                              color: Colors.white,
+                            ),
+                            onPressed: () {
+                              setState(() =>
+                                  AppModule.audioPlayerPageIsOffstage = true);
```
Comment (7flash) [inline lib/modules/audio_player/audio_player_page.dart:386]: ??

File: lib/modules/playlist/add_to_playlist/add_to_playlist_page.dart [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    tempInMemoryPlaylistCollection.remove(playlist);
+    tempInMemoryPlaylistCollection.add(updatedPlaylist);
+  }
+
+  void _close() {
+    if (widget.isPresentedAsOverlay) {
+      MediaOverlays.disposeAddToPlaylistOverlayEntry();
+      setState(() => AppModule.audioPlayerPageIsOffstage = false);
+    } else {
+      Modular.to.pop();
```
Comment (7flash) [inline lib/modules/playlist/add_to_playlist/add_to_playlist_page.dart:165]: Can we use Modular.to.maybePop() instead??

File: lib/modules/playlist/resources/assets.dart [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import 'package:flutter/material.dart';
+
+class PlaylistImageAssets {
+  static final Image newPlaylist = const Image(
```
Comment (7flash) [inline lib/modules/playlist/resources/assets.dart:4]: This is good idea to separate image routes into separate file, I will do the same with all images in separate PR.
https://gist.github.com/humazed/21dfe6a5371547f5a4e3543e80299011

File: lib/modules/playlist/resources/strings.dart [hunk 6]
```diff
@@ -0,0 +1,34 @@
+class PlaylistStrings {
```
Comment (7flash) [inline lib/modules/playlist/resources/strings.dart:1]: How would that work with multiple languages?

File: lib/modules/playlist/widgets/playlist_audio_item_tile.dart [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            icon: PlaylistImageAssets.closePlaylistAudioTile,
+            onPressed: () => onDelete(audioFile),
+          )
+        ],
+      ),
+    );
+  }
+
+  void _reorderAudiofilesInPlayList() {
+    print('_reorderAudiofilesInPlayList');
```
Comment (7flash) [inline lib/modules/playlist/widgets/playlist_audio_item_tile.dart:74]: why it does nothing?

File: lib/modules/playlist/widgets/playlist_more_menue.dart [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_modular/flutter_modular.dart';
+import 'package:highvibe/models/models.dart';
+import 'package:highvibe/modules/playlist/playlist_module.dart';
+import 'package:highvibe/modules/playlist/resources/resources.dart';
+import 'package:highvibe/widgets/header_row.dart';
+
+void playlistContextMenue(BuildContext context, PlayList playlist) {
```
Comment (7flash) [inline lib/modules/playlist/widgets/playlist_more_menue.dart:8]: menue - is that a typo?

File: lib/modules/playlist/widgets/playlist_more_menue.dart [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+          children: [
+            const HeaderRow(title: PlaylistStrings.options),
+            IconButton(
+              icon: PlaylistImageAssets.closeBottomSheet,
+              onPressed: () => Modular.to.pop(),
+            )
+          ],
+        ),
+        ListTile(
+          leading: PlaylistImageAssets.editBottonSheet,
```
Comment (7flash) [inline lib/modules/playlist/widgets/playlist_more_menue.dart:47]: botton is typo

Comment (7flash) [reply lib/modules/playlist/widgets/playlist_more_menue.dart:47]: button or bottom?

File: lib/modules/playlist/widgets/playlist_tile.dart [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+          )
+        ],
+      ),
+    );
+  }
+
+  int _calculateTotalDuration() {
+    var totalDurration = 0;
+    for (final item in widget.playList.audioFiles) {
+      totalDurration += item.duration;
```
Comment (7flash) [inline lib/modules/playlist/widgets/playlist_tile.dart:101]: durration is typo

File: lib/widgets/overlay_builder.dart [hunk 11]
```diff
@@ -0,0 +1,81 @@
+import 'package:flutter/material.dart';
+
+class OverlayBuilder extends StatefulWidget {
```
Comment (7flash) [inline lib/widgets/overlay_builder.dart:3]: where it comes from and where is going to be used?

**Comment by JasonHighVibe** [comment]
Hey guys; some notes regarding the playlist:

1. Create a new Playlist directly from the Add to Playlist Screen
2. Make the sweetspot a little larger to be able to reorder the tracks
3. The movement animation is a little choppy when the tracks are getting reordered
4. Get rid of Search on the Playlist Screen
5. Add a Small Saved Pop-up once a track has been added into the a playlist (will get this designed)
6. Clicking back from playlist menu on Native Android Navigation stopped working
7. Repeat track stopped working on the Audio Player screen.

File: lib/modules/playlist/resources/assets.dart (hunk 5 above)
Comment (minikin) [reply lib/modules/playlist/resources/assets.dart:4]: I'm not sure what's going on in the gist example.

File: lib/modules/playlist/resources/strings.dart (hunk 6 above)
Comment (minikin) [reply lib/modules/playlist/resources/strings.dart:1]: As I know we don't have support for localization and internalization for now. There are many ways how we can do it. When we'll decide to add  localization and internalization support string like e.g. 
`static const String addToPlaylist = 'Add to Playlist'` would be replaced with something like 
`static const String addToPlaylist = 'app.identifier.playlist.addaddToPlaylist.button.title'`

I can implement it, but I think it's out of scope for this pr.

File: lib/modules/audio_player/audio_player_page.dart (hunk 2 above)
Comment (minikin) [reply lib/modules/audio_player/audio_player_page.dart:32]: This app uses mobx + flutter_modular for state management.
These libs are deallocating widgets if they are not presented on screen. 
That's not the type of behavior that we need on `audio_player_page`.  
The easiest way to keep/control`audio_player_page` in memory while the `add_to_playlist_page` is presented is to wrap it in the [Offstage class](https://api.flutter.dev/flutter/widgets/Offstage-class.html).

File: lib/modules/audio_player/audio_player_page.dart (hunk 3 above)
Comment (minikin) [reply lib/modules/audio_player/audio_player_page.dart:386]: I would really appreciate it if you could find time to read [this](https://google.github.io/eng-practices/review/reviewer/). Thank you!

File: ios/Podfile.lock (hunk 1 above)
Comment (minikin) [reply ios/Podfile.lock:217]: [I don't agree ](https://guides.cocoapods.org/using/using-cocoapods.html#:~:text=Whether%20or%20not%20you%20check%20in%20the%20Pods%20directory%2C%20the,be%20kept%20under%20version%20control.)

**Comment by minikin** [comment]
@JasonHighVibe

1. Done.

2. I don't get it. Can you clarify it, pls? 

3. Have you checked it on real a device with an app compiled in [release mode](https://flutter.dev/docs/testing/build-modes)? It's a default animation. If we need to have some custom one, pls create a task.

4. I don't get it. Can you clarify it, pls? 

5. Ok. We can implement it in separate pr/branch when the design will be ready. Fire a task, pls.

6. I don't get it. Can you clarify it, pls? 

7. This PR is not related to Audio Player functionality. 
Only the `add to playlist` button was added on Audio Player screen. Feel free to fire a bug report.

8. Let's implement it in the next step `Playlist Firebase Integration`.

9. I can't reproduce it. Please fire a bug report. 

-----------

This task was too big, as a result, the PR review cycle is too long. 
I'd like to propose to do not have such big tasks in the future. We can just separate big tasks into several small ones.

Plus, to have a more detailed feature description, and thoroughgoing understanding of how a new feature is going to work with existing functionality can help a lot.

**Comment by JasonHighVibe** [comment]
@minikin 

Thanks for the feedback; I will make sure we breakdown the tasks into several smaller ones.
Will also introduce a more detailed feature description to make it easier for you as well.

To clarify the points I was addressing previously.

2. Make the sweetspot a little larger to be able to reorder the tracks
The point where you can use your finger to hold and drag to move the tracks seemed small and sometimes I would miss the sweetspot to move the track. I was saying it may be better to enlarge the sweetspot so it's easier to click and drag.

4. Get rid of search on the playlist screen
We can just leave this as is for now, but it's not neccesary to be able to search through all the playlists.

6. Clicking back from playlist menu on Native Android Navigation stopped working
When we were within the playlist menu, we clicked back on an Android phone and we were ejected from the App. @7flash can elaborate.

Other points I've covered with @7flash 

Thanks

**Comment by minikin** [comment]
@JasonHighVibe 

2. Do you mean on on this screen?
Screenshot 2020-06-19 at 10 20 44

**Comment by JasonHighVibe** [comment]
yes exactly

**Comment by minikin** [comment]
@JasonHighVibe 
We use Figma files as a reference for developing features.
What the reason, that Search is presented in UX/UI?

I'd like to propose to clean-up and update UX/UI before developers start work on features. 
It'll save time and keep good moral in the team.

ps: I think that it's a good idea to have a search for Playlists. Users can easily have 50 playlists or more.

**Comment by JasonHighVibe** [comment]
Sure; we're doing our best to make sure that UX/UI is ready before
implementation. Sometimes screens are getting changed as we go so sometimes
things will shift.

Agreed let's keep the Search then.

**Pull request merged**